compiler/rustc_trait_selection/src/error_reporting/infer/suggest.rs

```diff
@@ -476,6 +488,23 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         };
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        span: Span,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        terr: TypeError<'tcx>,
+        diag: &mut Diag<'_>,
+    ) {
+        debug!("suggest_function_pointers(cause={:?}, exp_found={:?})", cause, exp_found);
+
+        if exp_found.expected.peel_refs().is_fn() && exp_found.found.peel_refs().is_fn() {
+            self.suggest_function_pointers_impl(Some(span), exp_found, diag);
+        } else if let TypeError::Sorts(exp_found) = terr {
+            self.suggest_function_pointers_impl(None, &exp_found, diag);
```

Comment:
why not always provide the span here?

Reply:
because the span may not match the final exp_found

Reply:
I would expect the exp_found to still "contained" by that span, shouldn't it? as in, what's the impact on the diagnostic when always passing in the span?

Reply:
for the third case in the test, passing span would lead to:
```
help: consider casting to a fn pointer
   |
LL |     let _: Vec<fn()> = foo as fn();
   |                        ~~~~~~~~~~~
```